torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1501,9 +1501,12 @@ void InitXlaModuleBindings(py::module m) {
       // If the at::Tensor data is not present, we need to re-download the
       // tensor from the physical device to CPU. In that case, the value
       // must be present on the backend device.
-      XLA_CHECK(xtensor->GetXlaData() != nullptr &&
-                xtensor->CurrentDataHandle()->HasValue())
-          << "Cannot shard tensor. Data not present on any device.";
+      {
+        XLAGraphExecutor::Get()->DeviceBarrier(xtensor->GetDevice());
+        XLA_CHECK(xtensor->GetXlaData() != nullptr &&
+                  xtensor->CurrentDataHandle()->HasValue())
+            << "Cannot shard tensor. Data not present on any device.";
+      }
       std::vector<XLATensorPtr> xla_tensors{xtensor};
       cpu_tensor = XLAGraphExecutor::Get()->GetTensors(&xla_tensors)[0];
```

Comment:
Can we replace L1501 to 1511 with:
```
cpu_tensor = xtensor.ToTensor(true);
```

I tried but the figure out that it could break test_outbound_data_metrics. outbound_with_virtual_device will be equal to outbound_without_virtual_device.

Reply:
I'm not sure if `ToTensor` does what we need here.

Reply:
So what exactly do you need?

Reply:
I think we need to sync first why we are changing the API impl. This doesn't seem like enabling the tensor sharding attachment in the higher stack?

Reply:
Following up: `ToTensor` makes sense to me. Thanks Jiewen!

Reply:
Okay, I will leave that part to you to fix in the master, and I will remove the DeviceBarrier here.